### PR TITLE
New async connection and transaction API

### LINKED ISSUE #1540: Add async transaction methods
For 2.6.3 release we will add BeginTransactionAsync method, which will just open connection in async manner, if it is not opened, because ADO.NET doesn't have API for Begin/Commit/Rollback Transaction methods.

For next release we should try to improve this situation and implement async versions of those API for providers, that support it (at least mysqlconnector and npgsql).

Related issues with more information:
Async transaction API request for .net core: https://github.com/dotnet/corefx/issues/10975
Async transaction API request for ef core: https://github.com/aspnet/EntityFrameworkCore/issues/9015

### PR DESCRIPTION
**Pull request by MaceWindu**
Fixes #1540
- [x] add new async transaction API: CommitAsync, RollbackAsync, DisposeAsync
- [x] add new async connection API: CloseAsync, DisposeAsync, ChangeDatabaseAsync
- [x] add tests for new API
- [x] ~implement async version of new methods per-provider (only for providers that support them)~ implemented as reflection-based factory
- [x] add release/implementation notes
- [x] decide wether we need to implement methods, not used by linq2db now or not implemented yet by any provider:
- ~`IAsyncDisposable.DisposeAsync` for Connection - no implementors~ will wait for implementations
- ~`IAsyncDisposable.DisposeAsync` for Transaction - no implementors~ will wait for implementations
- ~Connection `ChangeDatabaseAsync`~ - not used by linq2db, so removed

Providers support status:

- Transaction `CommitAsync` : npgsql, mysqlconnector
- Transaction `RollbackAsync` : npgsql, mysqlconnector
- Connection `BeginTransactionAsync` : mysqlconnector, mysql.data (npgsql doesn't require this method)
- Connection `CloseAsync` : mysql.data
- Connection `ChangeDatabaseAsync` : mysqlconnector, mysql.data

Implementation notes:

With this PR we add `IAsyncDbConnection` and `IAsyncDbTransaction` abstractions over `IDbConnection` and `IDbTransaction` interfaces to use inside of linq2db. All instances of those abstractions must be created using `AsyncFactory.Create` methods. This factory already handles providers without async support and with support for methods, proposed by https://github.com/dotnet/corefx/issues/35012 (except `DisposeAsync`, also `Connection.CloseAsync` must return `Task`), so no changes to existing linq2db providers required.

For providers with `non-standard` async support, provider implementor must register corresponding factory methods using `AsyncFactory`.`RegisterConnectionFactory`/`RegisterTransactionFactory` methods

Other important changes:
- ~to support ValueTask type, linq2db adds `System.Threading.Tasks.Extensions` package dependency~ reverted, as it is not needed and brings dll-hell issues

### CODE REVIEW

**Comment by MaceWindu** [comment]
@sdanyliv , @it guys, please review proposed changes, then I will add actual implementation per provider if it's ok

File: Source/LinqToDB/Data/DataConnection.Async.cs [hunk 1]
```diff
@@ -88,6 +123,93 @@ public async Task EnsureConnectionAsync(CancellationToken cancellationToken = de
...
 
+		/// <summary>
+		/// Commits started (if any) transaction, associated with connection.
+		/// If underlying provider doesn't support asynchonous commit, it will be performed synchonously.
+		/// </summary>
+		/// <param name="cancellationToken">Asynchronous operation cancellation token.</param>
+		/// <returns>Asynchronous operation completion task.</returns>
+		public virtual async Task CommitTransactionAsync(CancellationToken cancellationToken = default)
+		{
+			if (TransactionAsync != null)
```
Comment (sdanyliv) [inline Source/LinqToDB/Data/DataConnection.Async.cs]: Something strange here with silent processing. Or we have to throw exception here or process command synchronously. This comment also for other methods below.

Comment (MaceWindu) [reply Source/LinqToDB/Data/DataConnection.Async.cs]: I think I need to clarify it: default implementations (AsyncDbConnection/transaction) will perform async calls synchonously. For specific providers I will override it with calls to async methods of underlying connection/transaction

**Comment by MaceWindu** [comment]
@sdanyliv , @igor-tkachev  - this PR is ready for review

**Comment by MaceWindu** [comment]
Release notes.

Added support for async transaction and connection close API:
```
// added by 2.6.4 and only updated to use provider's BeginTransactionAsync when awailable
Task<DataConnectionTransaction> DataConnection.BeginTransactionAsync(CancellationToken);
Task<DataConnectionTransaction> DataConnection.BeginTransactionAsync(IsolationLevel, CancellationToken);

// new API
Task DataConnection.CommitTransactionAsync(CancellationToken);
Task DataConnection.RollbackTransactionAsync(CancellationToken);

Task DataConnection.CloseAsync(CancellationToken);
Task DataConnection.DisposeAsync(CancellationToken);

Task DataConnectionTransaction.CommitAsync(CancellationToken);
Task DataConnectionTransaction.RollbackAsync(CancellationToken);

Task DataContextTransaction.CommitTransactionAsync(CancellationToken);
Task DataContextTransaction.RollbackTransactionAsync(CancellationToken);
```

Right now those API's are usefull only for following providers:
SQL Server: supports only `Connection.OpenAsync` part of `BeginTransactionAsync`
Npgsql: supports `CommitAsync`, `RollbackAsync`, `BeginTransactionAsync` (only open connection part as `BeginTransaction` is not blocking operation for npgsql)
MySqlConnector: supports `CommitAsync`, `RollbackAsync`, `BeginTransactionAsync`

Other providers doesn't contain async versions of required API (or implement them as fallback to synchonous operation) and new API calls for them will be performed synchronously.

### 3rd party providers support
Out-of-box `linq2db` will use async methods with following signatures:

###### Connection
Task<IDbTransactionBasedType> BeginTransactionAsync(CancellationToken);
Task<IDbTransactionBasedType> BeginTransactionAsync(IsolationLevel, CancellationToken);
Task CloseAsync(CancellationToken);

###### Transaction
Task CommitAsync(CancellationToken);
Task RollbackAsync(CancellationToken);

For providers that have required methods with other signature you need to implement your own `IAsyncDbConnection` or `IAsyncDbTransaction` instance factory and register in `AsyncFactory` class.

**Comment by MaceWindu** [comment]
Travis don't like C# 7.3 features. Will see if it can be enabled, otherwise will remove them

**Pull request merged**